[RFC] Inccomand substitution: Fix search highlight

KillTheMule opened the pull request:
Deactivate search highlighting dring incsub, and replace it by our own. 

Based on https://github.com/neovim/neovim/pull/7231.

Closes https://github.com/neovim/neovim/issues/5598.

KillTheMule committed:
Inccommand: Multiline substitutions, highlighting, multibyte.

Make inccomand work with multiline patterns and substitutions. Also care
for proper highlighting and multibyte characters.

KillTheMule committed:
Fix old inccomand tests, and add more for the new functionality.

KillTheMule committed:
Lint

KillTheMule committed:
Move tests into original file, lint, and add a test

KillTheMule committed:
Remove pat/sub from show_sub

They were only used to not show the preview window when typing "s/" or
"s//" only, in which case the previous pattern would be reused. Now the
window is shown in that case.

KillTheMule committed:
Fix highlighting conflict

KillTheMule committed:
Fix the last line in the preview buffer

It would always show an empty line at the end that didn't belong.

KillTheMule committed:
Adjust tests for the new preview window

... that does not have that superflous last line.

Also, remove some indeterminism for the freebsd64 tests. Partially,
those were suggested by the tests themselves, while successfull. Some of
them were added after some testing because the lookaround test would
fail on freebsd64 only.

KillTheMule committed:
Adjust tests for new highlighting.

Also extend an old test to show of the new way.

KillTheMule committed:
Remove superflous parameter from show_sub

justinmk commented on src/nvim/ex_cmds.c:
```diff
@@ -6156,7 +6249,11 @@ void ex_substitute(exarg_T *eap)
+  get_option_value((char_u *)"hls", &numval, NULL, 0);
+  set_option_value((const char *)"hls", 0, NULL, OPT_GLOBAL);
   buf_T *preview_buf = do_sub(eap, profile_setlimit(p_rdt));
+  set_option_value((const char *)"hls", numval, NULL, OPT_GLOBAL);
```
If possible, it's often faster to set/reset options directly (e.g. `p_hls` in this case), particuarly if there's side-effects. I usually look around the codebase to see how it's typically done for a given option (didn't check `p_hls`), some options will behave badly if their side-effects aren't taken care of. 

In this case I would _guess_ that `p_hls = false` is enough to get what we want, without any problems. Maybe you tried it already though.

KillTheMule replied:
I did neither try nor think of it :) I'll give it a a shot, but serious consideration will have to wait until the weekend.

KillTheMule commented:
I've changed setting `p_hls` to a direct call, and as far as I can see, that works, too. I've also added a little more testing, to show off how it's supposed to work :)

KillTheMule committed:
Clear highlight when there's no match

KillTheMule committed:
Test for clearing the highlight

justinmk commented:
@KillTheMule would you mind rebasing? I think we should just merge this.

KillTheMule committed:
Fix cmd modifier tests for the new highlight

KillTheMule commented:
Done.

KillTheMule commented:
There seems to be a failure in the oldtests on Travis, which did not happen before the rebase. I'll investigate tomorrow.

justinmk commented:
@KillTheMule that build failure is on master, don't worry about it.

justinmk commented on src/nvim/ex_cmds.c:
```diff
@@ -3999,22 +4050,30 @@ static buf_T *do_sub(exarg_T *eap, proftime_T timeout)
+        pre_src_id = bufhl_add_hl(NULL, 0, -1, 0, 0, 0);
+      }
+      if (pre_hl_id == 0) {
+        pre_hl_id = syn_check_group((char_u *)"Substitute", 13);
```
This should be `10`, right? I'll change it to `sizeof("Substitute") - 1`.

KillTheMule replied:
Ah yeah, thats a leftover from "IncSubstitute" I guess :)

justinmk commented:
@KillTheMule Nice work!
